pkg/types/symboltable/symboltable.go

```diff
@@ -0,0 +1,419 @@
+			return nil, err
+		}
+		symbol = &t
+	case gotypes.IdentifierType:
+		var t gotypes.Identifier
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	case gotypes.PointerType:
+		var t gotypes.Pointer
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	case gotypes.MethodType:
+		var t gotypes.Method
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		//FIXME: methods and hash-names - see prefix
+		symbol = &t
+	case gotypes.ChannelType:
+		var t gotypes.Channel
+		if err := json.Unmarshal(b, &t); err != nil {
+			return nil, err
+		}
+		symbol = &t
+	default:
+		return nil, fmt.Errorf("JSON unmarshal: Unknown symbol type '%v'", symbolType)
+	}
+
+	return symbol, nil
+}
+
+func (o *SymbolDef) UnmarshalJSON(b []byte) error {
+	var objMap map[string]*json.RawMessage
+	var symbol gotypes.DataType
+	var err error
+
+	if err = json.Unmarshal(b, &objMap); err != nil {
+		return err
+	}
+
+	if err = json.Unmarshal(*objMap["name"], &o.Name); err != nil {
+		return err
+	}
+
+	if symbol, err = unmarshalDataType(*objMap["def"]); err != nil {
+		return err
+	}
+
+	o.Def = symbol
+	return nil
+}
+
+///////////////////////////////////////////////////////////////////
+
+type Package struct {
+	//TODO: really?
+	variables map[string]*SymbolDef
+	functions map[string]*SymbolDef
+	datatypes map[string]*SymbolDef
+	constants map[string]*SymbolDef
+
+	Variables []*SymbolDef `json:"variables"`
+	DataTypes []*SymbolDef `json:"datatypes"`
+	Functions []*SymbolDef `json:"functions"`
+	Constants []*SymbolDef `json:"constants,omitempty"`
+	Name      string       `json:"package"`
+
+	//Project     string
+	//Imports     []string
+}
+
+func (o *Package) LookupSymbol(id string) *SymbolDef {
+	if sym, ok := o.variables[id]; ok {
+		return sym
+	} else if sym, ok := o.functions[id]; ok {
+		return sym
+	} else if sym, ok := o.datatypes[id]; ok {
+		return sym
+	} else if sym, ok := o.constants[id]; ok {
+		return sym
+	}
+
+	return nil
+}
+
+func (o *Package) Init() {
+	o.variables = make(map[string]*SymbolDef)
+	o.functions = make(map[string]*SymbolDef)
+	o.constants = make(map[string]*SymbolDef)
+	o.datatypes = make(map[string]*SymbolDef)
+
+	o.Variables = make([]*SymbolDef, 0)
+	o.Functions = make([]*SymbolDef, 0)
+	o.Constants = make([]*SymbolDef, 0)
+	o.DataTypes = make([]*SymbolDef, 0)
+}
+
+const errSymExists string = "Symbol '%s' already exists in symbol %T"
+
+//TODO: IT could be replaced by AddSymbol, but I don't know reliable way
+//      how to recognize each item for current structure and methods
+//      of SymbolDef; think about that
+
+func (o *Package) AddFunction(id string, sym *SymbolDef) error {
+	if _, ko := o.functions[id]; ko {
+		return fmt.Errorf(errSymExists, sym.Name, sym)
+	}
+
+	o.functions[id] = sym
+	o.Functions = append(o.Functions, sym)
+	return nil
+}
+
+func (o *Package) AddVariable(id string, sym *SymbolDef) error {
```

Comment:
The body of the function is almost the same as of the ``AddFunction``.  You could have:

```go
const ( 
        VariableSymbol = iota
        FunctionSymbol = iota
        DataTypeSymbol = iota
)

type Package struct {
   symbols map[int]map[string]*gotypes.SymbolDef
   Symbols map[int][]*gotypes.SymbolDef
}

func (o *Package) AddSymbol(symbolType int, id string, sym *SymbolDef) error {
        if _, ko := o.symbols[symbolType][id]; ko {
		return fmt.Errorf(errSymExists, sym.Name, sym)
	}

	o.symbols[symbolType][id] = sym
	o.Symbols[symbolType] = append(o.Symbols[symbolType], sym)
	return nil
}

func (o *Package) AddVariable(id string, sym *SymbolDef) error {
        return o.AddSymbol(VariableSymbol, id, sym)
}

func (o *Package) AddDataType(id string, sym *SymbolDef) error {
        return o.AddSymbol(DataTypeSymbol, id, sym)
}

...
```